zebra-consensus/src/transaction/check.rs

```diff
@@ -4,18 +4,49 @@
 
 use std::{borrow::Cow, collections::HashSet, convert::TryFrom, hash::Hash};
 
+use chrono::Utc;
+
 use zebra_chain::{
     amount::{Amount, NonNegative},
     block::Height,
     orchard::Flags,
     parameters::{Network, NetworkUpgrade},
     primitives::zcash_note_encryption,
     sapling::{Output, PerSpendAnchor, Spend},
-    transaction::Transaction,
+    transaction::{LockTime, Transaction},
 };
 
 use crate::error::TransactionError;
 
+/// Checks if the transaction's lock time allows this transaction to be included in a block.
+///
+/// Consensus rule:
+///
+/// > The transaction must be finalized: either its locktime must be in the past (or less
+/// > than or equal to the current block height), or all of its sequence numbers must be
+/// > 0xffffffff.
+///
+/// [`Transaction::lock_time`] validates the transparent input sequence numbers, returning [`None`]
+/// if they indicate that the transaction is finalized by them. Otherwise, this function validates
+/// if the lock time is in the past.
+pub fn lock_time_has_passed(tx: &Transaction, height: Height) -> Result<(), TransactionError> {
+    match tx.lock_time() {
+        Some(LockTime::Height(unlock_height)) => {
+            if height < unlock_height {
+                return Err(TransactionError::LockedUntilBlockHeight(unlock_height));
+            }
+        }
+        Some(LockTime::Time(unlock_time)) => {
+            if Utc::now() < unlock_time {
```

Comment:
Optional Refactor: We're more likely to get the rules correct if they're next to the code that implements them, and the code follows the rule.

```suggestion
            // > The transaction can be added to any block which has a greater height.
            // The Bitcoin documentation is wrong or outdated here,
            // so this code is based on the `zcashd` implementation at:
            // https://github.com/zcash/zcash/blob/1a7c2a3b04bcad6549be6d571bfdff8af9a2c814/src/main.cpp#L722
            if block_height > unlock_height {
                return Ok(());
            } else {
                return Err(TransactionError::LockedUntilBlockHeight(unlock_height));
            }
        }
        Some(LockTime::Time(unlock_time)) => {
            // > The transaction can be added to any block whose block time is greater than the locktime.
            // https://developer.bitcoin.org/devguide/transactions.html#locktime-and-sequence-number
            if block_time > unlock_time {
                return Ok(());
            } else {
```

Reply:
Done in [85df78b](https://github.com/jvff/zebra/commit/85df78b8d3c3de508460f89dfa36d6833467a6a2) and [944c684](https://github.com/jvff/zebra/commit/944c68466b8e7c55ad390077be48f07f8ddc9c67). There's also [dc6fabd](https://github.com/jvff/zebra/commit/dc6fabd03bddb418383786b44688450b098d1629)